### PR TITLE
Faster modulated pyramid wavefront sensor

### PR DESCRIPTION
**Pull request by syhaffert**
This is the implementation of the faster modulated pwfs as mentioned in #112 . I have added a test to check the new method versus the old version and they agree quite well with each other. The maximum relative error is smaller than 2 percent, and the relative rms of the difference is ~0.2 percent.

### CODE REVIEW

File: hcipy/wavefront_sensing/pyramid.py [hunk 1]
```diff
@@ -17,18 +18,43 @@ class ModulatedPyramidWavefrontSensorOptics(WavefrontSensorOptics):
 		The modulation radius in radians.
 	num_steps : int
 		The number of steps per modulation cycle.
+	fast_modulation_method : boolean
```
Comment (ehpor) [inline hcipy/wavefront_sensing/pyramid.py:21]: You should explain what the (dis)advantages of the fast modulation method are.

File: hcipy/wavefront_sensing/pyramid.py [hunk 2]
```diff
@@ -17,18 +18,43 @@ class ModulatedPyramidWavefrontSensorOptics(WavefrontSensorOptics):
...
+		separation = self.pyramid_wavefront_sensor._separation
+		refractive_index = self.pyramid_wavefront_sensor._refractive_index
+		wavelength_0 = self.pyramid_wavefront_sensor._wavelength_0
+
+		def surface_function(temp_grid):
+			return Field(-separation / (2 * (refractive_index(wavelength_0) - 1)) * (np.abs(temp_grid.x) + np.abs(temp_grid.y)), temp_grid)
+
+		self._pyramid_apodizers = []
+		for p in self.modulation_positions:
+			shifted_grid = self.focal_grid.shifted(2 * p)  # This factor two is to compensate for the 1/2 for the mirror surface.
```
Comment (ehpor) [inline hcipy/wavefront_sensing/pyramid.py:46]: I prefer to have this inline comment on a separate line. Use inline comments sparingly.

File: hcipy/wavefront_sensing/pyramid.py [hunk 3]
```diff
@@ -17,18 +18,43 @@ class ModulatedPyramidWavefrontSensorOptics(WavefrontSensorOptics):
...
+
+			pyramid_surface = evaluate_supersampled(surface_function, shifted_grid, 4)
+			pyramid_apodizer = SurfaceApodizer(Field(pyramid_surface, self.focal_grid), refractive_index)
+
+			self._pyramid_apodizers.append(pyramid_apodizer)
+
+		self.pupil_to_focal = self.pyramid_wavefront_sensor.pupil_to_focal
+		self.focal_to_pupil = self.pyramid_wavefront_sensor.focal_to_pupil
+
+		self._fast_modulation_method = fast_modulation_method
```
Comment (ehpor) [inline hcipy/wavefront_sensing/pyramid.py:56]: Move this to the top of `__init__()` with all the other parameter-to-variable copies.

File: hcipy/wavefront_sensing/pyramid.py [hunk 4]
```diff
@@ -42,13 +68,18 @@ def forward(self, wavefront):
...
 		'''
-		wf_modulated = []
+		if self._fast_modulation_method:
+			wf_foc = self.pyramid_wavefront_sensor.spatial_filter(self.pupil_to_focal(wavefront))
+			wf_modulated = [self.focal_to_pupil(apod(wf_foc)) for apod in self._pyramid_apodizers]
 
-		for point in self.modulation_positions.points:
-			self.tip_tilt_mirror.actuators = point
-			modulated_wavefront = self.tip_tilt_mirror.forward(wavefront)
+		else:
```
Comment (ehpor) [inline hcipy/wavefront_sensing/pyramid.py:75]: No empty line before the else.

File: tests/test_wavefront_sensing.py [hunk 5]
```diff
@@ -32,16 +32,27 @@ def test_modulated_pyramid_wavefront_sensor():
 
 	num_steps = 12
 
-	pywfs = PyramidWavefrontSensorOptics(pupil_grid, wfs_grid, 1.0)
+	pywfs = PyramidWavefrontSensorOptics(pupil_grid, wfs_grid, 1.0, q=3)
```
Comment (ehpor) [inline tests/test_wavefront_sensing.py:35]: Why this change? Because of systematic noise reasons of course, but does this mean that you _need_ to have higher `q` values for the fast modulation method? If so, that should definitely be in the docstring.

**Review by ehpor — CHANGES_REQUESTED** [review]
Just simple style/docstring changes, plus one clarification.

File: tests/test_wavefront_sensing.py (hunk 5 above)
Comment (syhaffert) [reply tests/test_wavefront_sensing.py:35]: The reason for this is that the code that chooses the correct focal plane sampling of the pyramid does not work correctly. I found that out last week. That is a bug that we need to fix. I am working on that in a separate issue. In the mean time I hard-coded the correct q for this simulation.

File: hcipy/wavefront_sensing/pyramid.py (hunk 2 above)
Comment (syhaffert) [reply hcipy/wavefront_sensing/pyramid.py:46]: done

File: hcipy/wavefront_sensing/pyramid.py (hunk 1 above)
Comment (syhaffert) [reply hcipy/wavefront_sensing/pyramid.py:21]: done

File: hcipy/wavefront_sensing/pyramid.py (hunk 4 above)
Comment (syhaffert) [reply hcipy/wavefront_sensing/pyramid.py:75]: done

File: hcipy/wavefront_sensing/pyramid.py (hunk 3 above)
Comment (syhaffert) [reply hcipy/wavefront_sensing/pyramid.py:56]: done

File: hcipy/wavefront_sensing/pyramid.py (hunk 1 above)
Comment (ehpor) [reply hcipy/wavefront_sensing/pyramid.py:21]: It's missing the disadvantages still. There need to be some, otherwise we would always use the fast method.

And I'd put this text in the main docstring rather than inside the parameters section.

Eg.:

```
'''The optical elements for a modulated pyramid wavefront sensor.

This class supports both a slow (default) and a fast modulation method, chosen by the
fast_modulation_method parameters. While the fast modulation method can be up to
twice as fast, it ....

Parameters
--------------
```

File: tests/test_wavefront_sensing.py (hunk 5 above)
Comment (ehpor) [reply tests/test_wavefront_sensing.py:35]: Okay.

**Review by ehpor — CHANGES_REQUESTED** [review]

File: hcipy/wavefront_sensing/pyramid.py (hunk 1 above)
Comment (syhaffert) [reply hcipy/wavefront_sensing/pyramid.py:26]: I think the only disadvantage is that it uses more memory. The two methods give really similar results so I don't know which one is actually more accurate compared to a true modulated pyramid. I went with:

`This class supports both a slow (default) and a fast modulation method, chosen by the
fast_modulation_method parameters. The fast method shifts the pyramid surface instead of
applying tilts to the incoming wavefront. While the fast modulation method can be up to
twice as fast, it needs to precompute all pyramid surfaces and will use up more memory.`

Comment (ehpor) [reply hcipy/wavefront_sensing/pyramid.py:26]: That sounds good. I just wanted something as the disadvantage. Otherwise there is no reason to not remove the slow method altogether.

**Comment by ehpor** [comment]
I'll approve once the tests have passed.

**Review by ehpor — APPROVED** [review]
LGTM.
